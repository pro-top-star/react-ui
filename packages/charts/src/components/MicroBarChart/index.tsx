import React, { CSSProperties, FC, forwardRef, Ref } from 'react';
import { createUseStyles, useTheme } from 'react-jss';
import { CommonProps } from '../../interfaces/CommonProps';
import { resolveColors } from '../../util/populateData';

const BarStyles = {
  container: {
    display: 'flex',
    flexDirection: 'column',
    overflow: 'hidden'
  },
  labelContainer: {
    display: 'flex',
    justifyContent: 'space-between'
  },
  valueContainer: {
    display: 'flex'
  },
  valueBar: { height: '4px' },
  fillUp: {
    height: '4px',
    backgroundColor: '#f1f2f3',
    flexGrow: 1
  },
  label: {
    color: 'hsl(0, 0%, 40%)',
    display: 'block',
    overflow: 'hidden',
    fontFamily: "'72', Arial, Helvetica, sans-serif",
    fontWeight: 'normal',
    whiteSpace: 'nowrap',
    textOverflow: 'ellipsis',
    fontSize: '12px',
    maxWidth: '70%'
  },
  text: {
    paddingLeft: '6px',
    display: 'inline-block',
    overflow: 'hidden',
    fontSize: '14px',
    boxSizing: 'border-box',
    fontFamily: "'72', Arial, Helvetica, sans-serif",
    fontWeight: 'normal',
    whiteSpace: 'nowrap',
    textOverflow: 'ellipsis',
    color: '#333333',
    textAlign: 'right'
  }
};

const useStyles = createUseStyles(BarStyles, { name: 'MicroBarChart' });

interface DataItems {
  value: number;
  label: string;
}

export interface MicroBarChartPropTypes extends CommonProps {
  dataset: DataItems[];
  maxWidth?: CSSProperties['width'];
  colors?: Array<CSSProperties['color']>;
  visibleDatasetCount?: number;
  valueFormatter?: (value: any) => string | number;
  labelFormatter?: (value: any) => string | number;
}

/**
 * <code>import { MicroBarChart } from '@ui5/webcomponents-react-charts/lib/MicroBarChart';</code>
 */
const MicroBarChart: FC<MicroBarChartPropTypes> = forwardRef(
  (props: MicroBarChartPropTypes, ref: Ref<HTMLDivElement>) => {
    const { className, dataset, colors, maxWidth, visibleDatasetCount, valueFormatter, labelFormatter, style } = props;
    const classes = useStyles();
    const theme: any = useTheme();
    const visibleDatasetArray = visibleDatasetCount ? dataset.slice(0, visibleDatasetCount) : dataset;

    const colorPalette = resolveColors(colors, theme.theme);

    const maxValue = Math.max(...dataset.map((item) => item.value));

    return (
      <div className={`${classes.container} ${className}`} style={{ maxWidth, ...style }} ref={ref}>
        {visibleDatasetArray.map((item, index) => {
          return (
            <div key={item.label}>
              <div className={classes.labelContainer}>
                <span className={classes.label} title={`${labelFormatter(item.label)}`}>
                  {labelFormatter(item.label)}
                </span>
                <span
                  className={classes.text}
                  style={{
                    fontSize: '12px'
                  }}
                  title={`${valueFormatter(item.value)}`}
                >
                  {valueFormatter(item.value)}
                </span>
              </div>
              <div className={classes.valueContainer}>
                <div
                  className={classes.valueBar}
                  style={{
                    width: `${(item.value / maxValue) * 100}%`,
                    backgroundColor: colorPalette[index]
                  }}
                />
                <div className={classes.fillUp} />
              </div>
            </div>
          );
        })}
      </div>
    );
  }
);

MicroBarChart.displayName = 'MicroBarChart';
MicroBarChart.defaultProps = {
  valueFormatter: (value) => value,
  labelFormatter: (value) => value
};

export { MicroBarChart };
