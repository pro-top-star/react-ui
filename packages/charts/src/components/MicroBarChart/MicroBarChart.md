```jsx
import { MicroBarChart } from '@ui5/webcomponents-react/lib/MicroBarChart';
```
