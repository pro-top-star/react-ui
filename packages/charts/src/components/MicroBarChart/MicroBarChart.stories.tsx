import React from 'react';
import { MicroBarChart } from './MicroBarChart';
import { ThemingParameters } from '@ui5/webcomponents-react-base';

const singleData = [
  {
    name: 'January',
    data: 100,
    users: 120
  },
  {
    name: 'February',
    data: 300,
    users: 40
  },
  {
    name: 'March',
    data: 530,
    users: 20
  },
  {
    name: 'April',
    data: 200,
    users: 30
  }
];

export default {
  title: 'Charts /  MicroBarChart',
  component: MicroBarChart,
  argTypes: {
    dataset: {
      type: null
    },
    measure: {
      type: null
    },
    dimension: {
      type: null
    }
  },
  args: {
    chartConfig: {
      yAxisVisible: false,
      xAxisVisible: true,
      gridStroke: ThemingParameters.sapList_BorderColor,
      gridHorizontal: true,
      gridVertical: false,
      legendPosition: 'bottom',
      legendHorizontalAlign: 'left',
      zoomingTool: false,
      resizeDebounce: 250
    }
  }
};

export const defaultStory = (props) => (
  <MicroBarChart
    onDataPointClick={props.onDataPointClickHandler}
    onLegendClick={props.onLegendClickHandler}
    dataset={singleData}
    style={{ width: '100%', height: '30vh' }}
    dimension={{
      accessor: 'name'
    }}
    measure={{
      accessor: 'users'
    }}
  />
);

defaultStory.storyName = 'Default';

export const withCustomColor = (props) => (
  <MicroBarChart
    onDataPointClick={props.onDataPointClickHandler}
    onLegendClick={props.onLegendClickHandler}
    dataset={singleData}
    style={{ width: '100%', height: '30vh' }}
    dimension={{
      accessor: 'name'
    }}
    measure={{
      accessor: 'users',
      colors: ['#f0ab00', 'var(--sapHighlightColor)'],
      formatter: (d) => (d > 200 ? 'over 200' : 'lower')
    }}
  />
);

withCustomColor.storyName = 'With custom color';

export const loadingPlaceholder = () => <MicroBarChart style={{ width: '20%' }} />;

loadingPlaceholder.storyName = 'Loading placeholder';
