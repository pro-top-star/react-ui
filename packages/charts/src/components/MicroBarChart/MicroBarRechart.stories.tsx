import { action } from '@storybook/addon-actions';
import React from 'react';
import { MicroBarChart } from './MicroBarChart';

const singleData = [
  {
    name: 'January',
    data: 100,
    users: 120
  },
  {
    name: 'February',
    data: 300,
    users: 40
  },
  {
    name: 'March',
    data: 530,
    users: 20
  },
  {
    name: 'April',
    data: 200,
    users: 30
  }
];

export default {
  title: 'Charts - Unstable /  MicroBarChart',
  component: MicroBarChart
};

export const defaultStory = () => (
  <MicroBarChart
    onDataPointClick={action('onDataPointClickHandler')}
    onLegendClick={action('onLegendClickHandler')}
    dataset={singleData}
    width={'100%'}
    height={'25vh'}
  />
);

defaultStory.story = {
  name: 'Default'
};

export const withCustomColor = () => (
  <MicroBarChart
    onDataPointClick={action('onDataPointClickHandler')}
    dataset={singleData}
    labelKey={'name'}
    width={'50%'}
    height={'25vh'}
    color={'red'}
    dataKeys={['users']}
  />
);

withCustomColor.story = {
  name: 'With custom color'
};

export const loadingPlaceholder = () => <MicroBarChart labelKey={'name'} width={'20%'} />;

loadingPlaceholder.story = {
  name: 'Loading placeholder'
};
