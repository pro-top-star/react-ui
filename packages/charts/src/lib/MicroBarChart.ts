import { MicroBarChart } from '../components/MicroBarChart';

export { MicroBarChart };
