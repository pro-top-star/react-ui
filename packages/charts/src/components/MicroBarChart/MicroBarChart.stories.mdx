import { Meta, Title, Description, Story, Canvas, ArgsTable } from '@storybook/addon-docs/blocks';
import { MicroBarChart } from '@ui5/webcomponents-react-charts/lib/MicroBarChart';

<Meta
  title="Charts /  MicroBarChart"
  component={MicroBarChart}
  argTypes={{
    dataset: { type: null },
    dimension: { type: null },
    measure: { type: null }
  }}
  args={{
    dataset: [
      {
        name: 'January',
        data: 100
      },
      {
        name: 'February',
        data: 300
      },
      {
        name: 'March',
        data: 530
      },
      {
        name: 'April',
        data: 200
      }
    ]
  }}
/>

<Title />
<Description />

<Canvas>
  <Story name="Default">
    {(args) => (
      <MicroBarChart
        dataset={args.dataset}
        loading={args.loading}
        style={{ width: '100%' }}
        dimension={{
          accessor: 'name'
        }}
        measure={{
          accessor: 'data'
        }}
      />
    )}
  </Story>
</Canvas>

<ArgsTable story="Default" of={MicroBarChart} />

<Canvas>
  <Story name="With custom props">
    {(args) => (
      <MicroBarChart
        onDataPointClick={args.onDataPointClick}
        dataset={args.dataset}
        style={{ width: '50%' }}
        loading={args.loading}
        dimension={{
          accessor: 'name'
        }}
        measure={{
          accessor: 'data',
          colors: ['#f0ab00', 'var(--sapHighlightColor)'],
          formatter: (d) => (d > 200 ? 'over 200' : 'lower'),
          width: 16
        }}
      />
    )}
  </Story>
</Canvas>

<Canvas>
  <Story name="Loading Placeholder">
    <MicroBarChart
      dataset={[]}
      dimension={{
        accessor: 'name'
      }}
      measure={{
        accessor: 'data'
      }}
      style={{ width: '20%' }}
    />
  </Story>
</Canvas>
