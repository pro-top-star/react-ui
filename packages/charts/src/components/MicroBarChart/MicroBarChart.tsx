import { enrichEventWithDetails } from '@ui5/webcomponents-react-base/lib/Utils';
import { ThemingParameters } from '@ui5/webcomponents-react-base/lib/ThemingParameters';
import { useConsolidatedRef } from '@ui5/webcomponents-react-base/lib/useConsolidatedRef';
import { ChartContainer } from '@ui5/webcomponents-react-charts/lib/next/ChartContainer';
import React, { CSSProperties, FC, forwardRef, Ref, useCallback, useMemo } from 'react';
import { Bar, BarChart as MicroBarChartLib, Cell, Tooltip, XAxis, YAxis } from 'recharts';
import { BarChartPlaceholder } from '../BarChart/Placeholder';
import { RechartBaseProps } from '../../interfaces/RechartBaseProps';
import { IChartMeasure } from '../../interfaces/IChartMeasure';

interface MeasureConfig extends Omit<IChartMeasure, 'accessor' | 'color'> {
  /**
   * A string containing the path to the dataset key this bar should display.
   * Supports object structures by using `'parent.child'`. Can also be a getter.
   */
  accessor: string;
  /**
   * array of any valid CSS Color or CSS Variable. Defaults to the `sapChart_OrderedColor_` colors
   */
  colors?: CSSProperties['color'][];
  /**
   * Bar Opacity
   * @default 1
   */
  opacity?: number;
  /**
   * Bar Width
   * @default auto
   */
  width?: number;
}

interface DimensionConfig {
  /**
   * A string containing the path to the dataset key this bar should display.
   * Supports object structures by using `'parent.child'`. Can also be a getter.
   */
  accessor: string;
  /**
   * function will be called for each data label and allows you to format it according to your needs
   */
  formatter?: (value: any) => string;
}

export interface MicroBarChartProps extends RechartBaseProps {
  centerLabel?: string;
  dimension: DimensionConfig;
  /**
   * An array of config objects. Each object is defining one bar in the chart.
   *
   * <h4>Required properties</h4>
   * - `accessor`: string containing the path to the dataset key this bar should display. Supports object structures by using <code>'parent.child'</code>.
   *
   * <h4>Optional properties</h4>
   *
   * - `formatter`: function will be called for each data label and allows you to format it according to your needs
   * - `colors`: array of any valid CSS Color or CSS Variable. Defaults to the `sapChart_OrderedColor_` colors
   */
  measure: MeasureConfig;
}

const TiltedAxisTick = (props) => {
  const { x, y, payload } = props;
  return (
    <g transform={`translate(${x},${y})`}>
      <text x={7} y={-10} textAnchor="begin" fill={ThemingParameters.sapContent_LabelColor}>
        {payload.value}
      </text>
    </g>
  );
};

/**
 * <code>import { MicroBarChart } from '@ui5/webcomponents-react-charts/lib/next/MicroBarChart';</code>
 * **This component is under active development. The API is not stable yet and might change without further notice.**
 */
const MicroBarChart: FC<MicroBarChartProps> = forwardRef((props: MicroBarChartProps, ref: Ref<any>) => {
  const { loading, dataset, onDataPointClick, style, className, tooltip, slot } = props;

  const chartRef = useConsolidatedRef<any>(ref);

  const dimension: DimensionConfig = useMemo(
    () => ({
      formatter: (d) => d,
      ...props.dimension
    }),
    [props.dimension]
  );
  const measure: MeasureConfig = useMemo(
    () => ({
      formatter: (d) => d,
      ...props.measure
    }),
    [props.measure]
  );

  const tooltipValueFormatter = useCallback((value) => measure.formatter(value), [measure.formatter]);

  const onDataPointClickInternal = useCallback(
    (e, i) => {
      if (e && onDataPointClick) {
        onDataPointClick(
          enrichEventWithDetails(e, {
            dataKey: Object.keys(e).filter((key) =>
              e.value.length ? e[key] === e.value[1] - e.value[0] : e[key] === e.value && key !== 'value'
            )[0],
            value: e.value.length ? e.value[1] - e.value[0] : e.value,
            payload: e.payload,
            xIndex: i
          })
        );
      }
    },
    [onDataPointClick]
  );

  const CustomizedLabel = (properties) => {
    const { x, y, value } = properties;
    const xText = chartRef.current.clientWidth - 60;

    return (
      <g transform={`translate(${x},${y})`}>
        <text y={-7} dx={`${xText}px`} textAnchor={'end'} fill={ThemingParameters.sapContent_LabelColor}>
          {measure.formatter(value)}
        </text>
      </g>
    );
  };

  return (
    <ChartContainer
      dataset={dataset}
      loading={loading}
      Placeholder={BarChartPlaceholder}
      ref={chartRef}
      style={style}
      className={className}
      tooltip={tooltip}
      slot={slot}
    >
      <MicroBarChartLib
        margin={{ left: -30, right: 30, top: 40, bottom: 30 }}
        layout={'vertical'}
        data={dataset}
        label={{ position: 'insideBottomRight', fill: ThemingParameters.sapContent_LabelColor }}
      >
        <XAxis hide type="number" />
        <YAxis
          axisLine={false}
          tick={<TiltedAxisTick />}
          tickLine={false}
          type="category"
          dataKey={dimension.accessor}
        />
        <Bar
          background={{ fill: ThemingParameters.sapNeutralBackground }}
          strokeOpacity={measure.opacity}
          fillOpacity={measure.opacity}
          label={{ content: <CustomizedLabel external={style.width} /> }}
          dataKey={measure.accessor}
          barSize={measure.width ?? 5}
          onClick={onDataPointClickInternal}
        >
          {dataset &&
            dataset.map((data, index) => (
              <Cell
                key={index}
                name={dimension.formatter(data[dimension.accessor])}
                fill={measure.colors?.[index] ?? `var(--sapChart_OrderedColor_${(index % 11) + 1})`}
              />
            ))}
        </Bar>
        <Tooltip cursor={{ fillOpacity: 0.3 }} formatter={tooltipValueFormatter} />
      </MicroBarChartLib>
    </ChartContainer>
  );
});

MicroBarChart.displayName = 'MicroBarChart';

export { MicroBarChart };
