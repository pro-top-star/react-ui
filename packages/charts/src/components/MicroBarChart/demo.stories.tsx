import { array, number, text } from '@storybook/addon-knobs';
import { MicroBarChart } from '@ui5/webcomponents-react-charts/lib/MicroBarChart';
import React from 'react';
import notes from './MicroBarChart.md';

const dataset = [
  { value: 10, label: 'Bar Number One' },
  { value: 100, label: 'Bar Number Two' },
  { value: 30, label: 'Bar Number Three' },
  { value: 40, label: 'Bar Number Four' },
  { value: 50, label: 'Bar Number Five' }
];
const colors = ['sapUiChartBad', 'pink', '#de890d', 'rgb(33, 9, 205)', 'green'];
const valueFormatter = (value) => `${value}$`;
const labelFormatter = (value) => `${value} in Dollar`;

export default {
  title: 'Charts / Micro Bar Chart',
  component: MicroBarChart,
  parameters: { notes }
};

export const defaultStory = () => <MicroBarChart dataset={dataset} />;

defaultStory.story = {
  name: 'Default'
};

export const withProps = () => (
  <MicroBarChart
    dataset={dataset}
    maxWidth={text('maxWidth', '200px')}
    colors={array('colors', colors)}
    visibleDatasetCount={number('visibleDatasetCount', 5)}
    valueFormatter={valueFormatter}
    labelFormatter={labelFormatter}
  />
);

withProps.story = {
  name: 'With Props'
};
