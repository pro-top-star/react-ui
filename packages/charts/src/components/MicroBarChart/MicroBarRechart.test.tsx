import { mount } from 'enzyme';
import * as React from 'react';
import { singleData } from '../../resources/DemoProps';
import { MicroBarChart } from './MicroBarChart';

describe('PieChart', () => {
  test('Renders with data', () => {
    expect(
      mount(
        <MicroBarChart
          dataset={singleData}
          style={{ width: '50%' }}
          dimension={{
            accessor: 'name'
          }}
          measure={{
            accessor: 'users'
          }}
        />
      ).render()
    ).toMatchSnapshot();
  });

  test('loading placeholder', () => {
    const wrapper = mount(<MicroBarChart style={{ width: '50%' }} dimensions={[]} measures={[]} />);
    expect(wrapper.render()).toMatchSnapshot();
  });
});
