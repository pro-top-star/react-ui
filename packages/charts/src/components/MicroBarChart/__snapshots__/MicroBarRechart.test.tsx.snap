// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PieChart Renders with data 1`] = `
<div
  style="font-family: var(--sapFontFamily); width: 50%; height: 400px; position: relative;"
>
  <div
    class="recharts-responsive-container"
    id="undefined"
    style="width: 100%; height: 100%;"
  >
    <div />
  </div>
</div>
`;

exports[`PieChart loading placeholder 1`] = `
<div
  style="font-family: var(--sapFontFamily); width: 50%; height: 400px; position: relative;"
>
  <svg
    aria-labelledby="ARIA-LABELLED-BY"
    height="100%"
    preserveAspectRatio="xMidYMid meet"
    role="img"
    viewBox="0 0 165 145"
    width="100%"
  >
    <title
      id="TITLE_ARIA_ID"
    >
      Loading...
    </title>
    <rect
      clip-path="CLIP-PATH-URL"
      height="100%"
      role="presentation"
      style="fill: url(#STYLE-URL);"
      width="100%"
      x="0"
      y="0"
    />
    <defs
      role="presentation"
    >
      <clipPath
        id="CLIP-PATH-URL"
      >
        <rect
          height="135"
          width="1"
          x="20"
          y="10"
        />
        <rect
          height="15"
          width="85"
          x="20"
          y="20"
        />
        <rect
          height="15"
          width="117"
          x="20"
          y="40"
        />
        <rect
          height="15"
          width="67"
          x="20"
          y="60"
        />
        <rect
          height="15"
          width="75"
          x="20"
          y="80"
        />
        <rect
          height="15"
          width="125"
          x="20"
          y="100"
        />
        <rect
          height="15"
          width="60"
          x="20"
          y="120"
        />
      </clipPath>
      <linearGradient
        id="STYLE-URL"
      >
        <stop
          offset="0%"
          stop-color="var(--sapContent_ImagePlaceholderBackground)"
          stop-opacity="var(--sapContent_DisabledOpacity)"
        >
          <animate
            attributeName="offset"
            dur="2s"
            keyTimes="0; 0.25; 1"
            repeatCount="indefinite"
            values="-2; -2; 1"
          />
        </stop>
        <stop
          offset="50%"
          stop-color="var(--sapContent_ImagePlaceholderForegroundColor)"
          stop-opacity="1"
        >
          <animate
            attributeName="offset"
            dur="2s"
            keyTimes="0; 0.25; 1"
            repeatCount="indefinite"
            values="-1; -1; 2"
          />
        </stop>
        <stop
          offset="100%"
          stop-color="var(--sapContent_ImagePlaceholderBackground)"
          stop-opacity="var(--sapContent_DisabledOpacity)"
        >
          <animate
            attributeName="offset"
            dur="2s"
            keyTimes="0; 0.25; 1"
            repeatCount="indefinite"
            values="0; 0; 3"
          />
        </stop>
      </linearGradient>
    </defs>
  </svg>
</div>
`;
